styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -110,53 +124,76 @@ private void triggerBackfill(Backfill backfill, Map<String, Long> backfillStates
 
     final Workflow workflow = workflowOpt.get();
 
+    while (true) {
+      try {
+        if (!storage.runInTransaction(tx -> tryTriggerAndProgress(tx, backfill.id(), workflow))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("transaction failure when working on backfill {}", backfill, e);
+        break;
+      }
+    }
+  }
+
+  private Boolean tryTriggerAndProgress(StorageTransaction tx,
+                                        String id, Workflow workflow) {
+    final Backfill momentBackfill =
+        storage.backfill(id).orElseThrow(RuntimeException::new);
+
+    final Instant partition = momentBackfill.nextTrigger();
     final int remainingCapacity =
-        backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+        momentBackfill.concurrency() - stateManager.activeStates(id).size();
 
-    Instant partition = backfill.nextTrigger();
+    if (remainingCapacity < 1) {
+      LOG.debug("no capacity left for backfill {}, stop triggering", momentBackfill);
```

Comment:
Should we store again the backfill with `.allTriggered(true)` just in case something went wrong when storing that after the last trigger in a previous call to `tryTriggerAndProgress`?

Reply:
Maybe to be paranoid. :)

The last storing (next trigger and that flag) all happens within a transaction. Not sure how that would happen.

Reply:
> The last storing (next trigger and that flag) all happens within a transaction. Not sure how that would happen.

Good point, that's enough